POEM_052.md

```diff
@@ -16,6 +16,112 @@ Associated implementation PR:
+Does OpenMDAO run the compute to test things during setup? In other words, does this skeleton  need to return something?
+
+Next steps include writing the function code and configuring the Component.
+Component configuration is what's included in the I/O and "everything else".
+For the I/O, I will ignore the fully de-coupled engineering code/ OpenMDAO API case.
+For the coupled case, we can do more annotations or decorators.
+
+For the annotations, it could be nice to provide a custom type that takes the metadata fields as var. And read defaults as default value like:
+
+```python
+def some_component_func(x: om.func.Input(units='m') = 0., y, z) ->
+                        [om.func.Output('foo'),
+                         om.func.Output('bar', units='1/m', shape=4)]:
+    return
+
+SomeComponent = om.ExplicitFuncComp(some_component_func)
+```
+Or if it's better to use a dict or 3rd party metadata container, that's fine too. Or support both. 
+
+Or, the input/output can be configured by decorator as proposed by Justin. I would like it if the input decorators inspected the name of the input variable by default. 
+
+For more involved OpenMDAO API calls, we can use a subclass to access the the full API. See below.
+Another feature is other functions having a normal signature, either as stand-alone functions or methods of subclass. Could use standard OpenMDAO method name and decorate to handle function input/output. For standalone, may need special API call or flag.
+Or are both cases solved by just assuming that signature type for this subclass and wrapping everything with an IO handler.
+
+```python
+class SomeComponent(om.ExplicitFuncComp(some_component_func)):
+    def setup(self):
+        super().setup()
+        """
+        can make any any component api calls here. If desired, could provide hooks (e.g., pre/post setup)
+        could do something similar with initialize, or any other method.
+
+        Is it possible to manipulate a 
+        """
+
+     @om.func.process_io # or maybe this is the default behavior for this class?
+     def compute_partials(self, x, y, z, J): 
+
+        J['foo', 'x'] = -3*np.log(z)(3*x+2*y)**2 
+        J['foo', 'y'] = -2*np.log(z)(3*x+2*y)**2 
+        J['foo', 'z'] = 1/(3*x+2*y) * 1/z
+
+        J['bar', 'x'][:] = 2 # need to set all elements of array
+        J['bar', 'y'][:] = 1 # need to set all elements of a
+
+```
+
+```python
+def J_some_func(x, y, z, J): 
+
+    J['foo', 'x'] = -3*np.log(z)(3*x+2*y)**2 
+    J['foo', 'y'] = -2*np.log(z)(3*x+2*y)**2 
+    J['foo', 'z'] = 1/(3*x+2*y) * 1/z
+
+    J['bar', 'x'][:] = 2 # need to set all elements of array
+    J['bar', 'y'][:] = 1 # need to set all elements of a
+
+class SomeComponent(om.ExplicitFuncComp(some_component_func)):
```

Comment:
I hate this slightly less... but I don't understand why you want to mingle the class and function based API's this way. 
Why not just pass the name of the derivative method as an additional argument to the FuncComp constructor? That seems much much cleaner than sub-classing! 

Also, the same *instance* vs *class* problem still shows up here.